### ISSUE TITLE
"Invalid identity pool configuration. Check assigned IAM roles for this pool."

### ISSUE THREAD

**Issue by Jahans3** [aws-amplify/amplify-js#6195]
**Describe the bug**
It appears the AWS SDK is not being configured correctly.

I was adding storage to my app but I kept getting 403 errors every time I tried to upload. After some digging I tried calling `Auth.currentUserCredentials`, which returns the following error:

```
InvalidIdentityPoolConfigurationException: Invalid identity pool configuration. Check assigned IAM roles for this pool.
```

In the console I have checked the IAM roles and user pool and they are configured correctly, so I assume the issue lies in how the config is being handled.

**To Reproduce**

1. `amplify init`
2. `amplify add auth`
3. `amplify add storage` (allow authenticated users to read, create and update. Unauthenticated users allowed to read only)
4. Create a user
5. Log in as user
6. Attempt to upload an image with `Storage.put`
7. See 403 error
8. Call `Auth.currentUserCredentials`
9. See `InvalidIdentityPoolConfigurationException` error

**Expected behavior**
I should be able to authenticate and upload files using the AWS SDK as documented in the official documentation: https://docs.amplify.aws/lib/storage/upload/q/platform/js

**Code Snippet**
Inside any handler:
```js
onPress={() => Storage.put('user/avatar', blob)}
```

Inside any component:
```js
  React.useEffect(() => {
    Auth.currentUserCredentials().then(console.log);
  }, []);
```

**Comment by Jahans3** [comment]
I've confirmed it's an issue with credentials not being properly applied by downgrading to 2.1.1, where I received a different error:

```
[WARN] 15:35.575 AWSS3Provider - error uploading CredentialsError: Missing credentials in config
```

**Comment by Amplifiyer** [comment]
@Jahans3 just to confirm, you did run amplify push after adding auth and storage resources? If yes, can you paste your aws_exports.js file generated. (Obfuscate any sensitive information as needed).

**Comment by Jahans3** [comment]
> @Jahans3 just to confirm, you did run amplify push after adding auth and storage resources?

Yes, of course

Here's my aws-exports:

```js
/* eslint-disable */
// WARNING: DO NOT EDIT. This file is automatically generated by AWS Amplify. It will be overwritten.

const awsmobile = {
    "aws_project_region": "eu-west-2",
    "aws_appsync_graphqlEndpoint": "https://<redacted>.appsync-api.eu-west-2.amazonaws.com/graphql",
    "aws_appsync_region": "eu-west-2",
    "aws_appsync_authenticationType": "API_KEY",
    "aws_appsync_apiKey": "<redacted>",
    "aws_cognito_identity_pool_id": "eu-west-2:<redacted>",
    "aws_cognito_region": "eu-west-2",
    "aws_user_pools_id": "eu-west-2_<redacted>",
    "aws_user_pools_web_client_id": "<redacted>",
    "oauth": {},
    "aws_user_files_s3_bucket": "<redacted>",
    "aws_user_files_s3_bucket_region": "eu-west-2"
};


export default awsmobile;

```

**Comment by Amplifiyer** [comment]
thanks @Jahans3, it looks related to https://github.com/aws-amplify/amplify-js/issues/3744
Check the amplify imports you are including, avoiding `@aws-amplify/core` and instead using 'aws-amplify' or check other solutions mentioned in the related issue. Let us know if one works.

**Comment by Jahans3** [comment]
Thanks @Amplifiyer 

I can confirm it's an issue with how `Amplify.configure` handles the `aws-exports.js` file. Hopefully we can get this issue resolved before long (I'm happy to help), will also save other people time if this is mentioned in the docs while a fix is being worked on.

I was able to workaround the issue by manually applying the properties from `aws-exports.js` to `Amplify.configure`:

```js
Amplify.configure({
  Auth: {
    identityPoolId: aws_exports['aws_cognito_identity_pool_id'], //REQUIRED - Amazon Cognito Identity Pool ID
    region: aws_exports['aws_project_region'], // REQUIRED - Amazon Cognito Region
    userPoolId: aws_exports['aws_user_pools_id'], //OPTIONAL - Amazon Cognito User Pool ID
    userPoolWebClientId: aws_exports['aws_user_pools_web_client_id'] //OPTIONAL - Amazon Cognito Web Client ID
  },
  Storage: {
    AWSS3: {
      bucket: aws_exports['aws_user_files_s3_bucket'], //REQUIRED -  Amazon S3 bucket
      region: aws_exports['aws_user_files_s3_bucket_region'] //OPTIONAL -  Amazon service region
    }
  }
});
```

**Comment by Amplifiyer** [comment]
@Jahans3 , is this instead of calling Amplify.configure() or are you calling Amplify.configure(manual overriden configuration) elsewhere, e.g. where you call Storage APIs?

**Comment by Jahans3** [comment]
@Amplifiyer this is instead of calling `Amplify.configure`. It also works like this too:

```js
Amplify.configure(aws_exports);
Amplify.configure({
  Storage: {
    AWSS3: {
      bucket: aws_exports['aws_user_files_s3_bucket'], //REQUIRED -  Amazon S3 bucket
      region: aws_exports['aws_user_files_s3_bucket_region'] //OPTIONAL -  Amazon service region
    }
  }
});
```

**Comment by lisandrolan** [comment]
I have the same issue and changing the export file does not solve it.

**Comment by ashika01** [comment]
@lisandrolan are you setting up exactly as mentioned here as well? https://github.com/aws-amplify/amplify-js/issues/6195#issue-647362942

**Comment by lisandrolan** [comment]
my use case is different, Im trying to login with tokens and I cannot get the session.

**Comment by sammartinez** [comment]
@Jahans3 Please let us know if you are still experiencing this issue with the latest version of Amplify and we can dig in.

**Comment by timoteialbu** [comment]
I just upgraded to the latest version of these packages:
```
"aws-amplify": "3.3.4"
"aws-amplify-react": "4.2.8"
```
and the issue persists.

**Comment by timoteialbu** [comment]
I found a way to solve my issue in regards to the `InvalidIdentityPoolConfigurationException: Invalid identity pool configuration. Check assigned IAM roles for this pool.` error.

The solution was to look up the `authRole` that amplify-cli creates and add trust relationships to it. Here is what I edited it to:

```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Federated": "cognito-identity.amazonaws.com"
      },
      "Action": "sts:AssumeRoleWithWebIdentity",
      "Condition": {
        "StringEquals": {
          "cognito-identity.amazonaws.com:aud": "us-west-2:<COGNITO_IDENTITY_POOL_ID>"
        },
        "ForAnyValue:StringLike": {
          "cognito-identity.amazonaws.com:amr": "authenticated"
        }
      }
    }
  ]
}
```

Since this was working prior to June 2020 (which was the last time we successfully created a new env and had it working with no extra changes like this), what has changed since? I looked through the [changelog](https://github.com/aws-amplify/amplify-cli/blob/master/packages/amplify-cli/CHANGELOG.md) but could not find anything.

Are we supposed to add trust relationships ourselves?

**Comment by Jahans3** [comment]
> @Jahans3 Please let us know if you are still experiencing this issue with the latest version of Amplify and we can dig in.

Hey @sammartinez - apologies for the slow response!

I have updated to the latest version and reverted the config to the standard `Amplify.configure(awsExports)` and am happy to confirm it's all working as expected now.

Will close this issue, thanks @Amplifiyer @sammartinez

**Comment by koterthecoder** [comment]
@timoteialbu Is your solution of manually changing the trust relationship required to make it work?

I'm encountering this issue with my production environment (dev environment works fine). Even after upgrading packages I'm still encountering the issue. 

```
    "aws-amplify": "^3.3.11",
    "aws-amplify-react": "^4.2.15",
```

I'm able to authenticate my user just fine. But when trying to upload a file using Storage.put, i'm getting four 400s from Cognito, then getting a 403 on the put request. I'm noticing that the error message it says `The AWS Access Key Id you provided does not exist in our records.` But I assume that is because of the initial bad requests against Cognito.

![image](https://user-images.githubusercontent.com/42450792/101263280-14a18200-3712-11eb-9566-7980729b3dfb.png)

**Comment by DavidBrear** [comment]
I ran into this error but it wasn't to do with the `authRole` as listed above. Mine was because I had manually created a group on Cognito and then was using that group and its associated role with some users. I wanted to document what I did in case anyone gets to this point and isn't able to solve the problem.

**TL;DR: Don't manually create groups and roles if you're going to be using them for Amplify.**

What I did to find out what the role that was causing the issue was to take the `access_token` that gets returned on the `POST` request to AWS. You can take this token and plug it into https://jwt.io/ which will decode it.

My issue was that any account where I was getting the above error included the following attributes in that JWT
```
"cognito:roles": [
    "arn:aws:iam::*ACCOUNT_ID*:role/*THE_ROLE*"
  ],
"cognito:preferred_role": "arn:aws:iam::*ACCOUNT_ID*:role/*THE_ROLE*",
```

If you are getting this issue, find these roles and verify that they have the conditions attached to them from @timoteialbu's response: ```
"Condition": {
        "StringEquals": {
          "cognito-identity.amazonaws.com:aud": "us-west-2:<COGNITO_IDENTITY_POOL_ID>"
        },
        "ForAnyValue:StringLike": {
          "cognito-identity.amazonaws.com:amr": "authenticated"
        }
      }```

My solution was to just manually delete the group I'd created along with the role associated with it and then recreate it via the Amplify CLI and then just move all the users who were in the original group into the new one. This way Amplify maintains it when it changes.
